lib/c/arm8_intrinsics.h

```diff
@@ -0,0 +1,16 @@
+
+static inline uint64_t
+uint128_t_add_carry_u64(uint64_t cin, uint64_t x, uint64_t y, uint64_t* r) {
+  __uint128_t res = (__uint128_t)x + cin + y;
+  r[0U] = (uint64_t)res;
+  uint64_t c = res >> 64;
+  return c;
+}
+
+static inline uint64_t
+uint128_t_sub_borrow_u64(uint64_t cin, uint64_t x, uint64_t y, uint64_t* r) {
+  __uint128_t res = (__uint128_t)x - cin - y;
```

Comment:
Do we assume `x > y + cin` or is this checked anywhere?

Reply:
No, but it's expected that `x` can be less than `y + cin` and in that case a borrow is returned, right?

Reply:
Maybe. I the functions need at least a comment with a contract what exactly it is supposed to do.
Looking at the IntTypes implementation this appears to be the case. But I think we don't cover one case `FStar_UInt64_eq_mask(res, x) & cin`, i.e. when the result is the same as x (smaller than 2^64) and we had a carry in. In this case this function here wouldn't return a carry but should.

Reply:
I don't think we're missing this case. Correct me if I'm wrong but the way I interpret `Hacl_IntTypes_Intrinsics_sub_borrow_u64` is the following:
* if res < x then c = 0
* if res > x then c = 1
* if res == x then either
  - y == 0 and cin == 0, so c = 0
  - y == UINT64_MAX and cin == 1, so c = 1
I don't see how res could be equal to x in any other way. If that's the case then I think the function here does correctly a carry in this situation.

Reply:
hm, I think you're right 😊

Reply:
you can verify that code directly in F*/Low* :) It supports 64-bit and 128-bit integers.

Reply:
good point Marina, that's actually how the "fallback" implementation works... that would be an excellent intern project, see https://github.com/project-everest/hacl-star/tree/master/code/fallback/

Reply:
It might be a good idea to drop that fallback implementation in favor of a new one if the performance numbers on intel are also better.